docs: add more information about blueprints and entities (#17)
equinor/dm-docs: docs/docs/concepts/blueprints.md

---
sidebar_position: 1
---

# Blueprints and entities

> Blueprints can be used to create domain models. 

Reviewer: ```suggestion
Write blueprints to define your domain models 
```
Is this like a tagline?

:::info

A blueprint is a model and an instance of a blueprint is called an entity. 

:::

## What is a Blueprint?

> A blueprint is described in [JSON](https://www.json.org/json-en.html). 

Reviewer: ```suggestion
Blueprints are files written in the Data Modelling schema, describing how entities created from that blueprint, should look like.

A blueprint is described in [JSON](https://www.json.org/json-en.html). 
```

Below is a short version of a Car blueprint. 

```json
{
  "name": "Car",

>   "type": "dmss://system/SIMOS/Blueprint",

Reviewer 2: we discussed in the meeting today (6.1) that the car blueprint should be a bit more advanced to showcase more features. I am creating a new issue for this.

Reviewer 2: https://github.com/equinor/dm-docs/issues/19

  "description": "This describes how a car looks like",
  "attributes": [
    {
      "attributeType": "string",
      "type": "dmss://system/SIMOS/BlueprintAttribute",
      "name": "name"
    },
    {
      "attributeType": "string",
      "type": "dmss://system/SIMOS/BlueprintAttribute",
      "name": "description"
    }
  ]
}
```

> Any blueprint can be understood by simply looking at it. 

Reviewer: ```suggestion
```

By looking at the attributes `name`, `type` and `attributes` in the example above, you can find out that:

* It represents a Car model 

> * It is of type `system/SIMOS/Blueprint`

Reviewer: Skip this for now?
Could be confusing if you dont explain that Blueprints are also entities of the master Blueprint.

Reviewer 3: An explanation of that would be nice. I at least found that a bit confusing until Eirik Ola explainit it to me

Author: Lets just keep, we might do a refactor of docs.

* It has the two attributes `name` and `description`.

> ## What is an entity?

Reviewer: Should get some diagrams in here. Showing the relationship of blueprints and entities.
Could be done in another PR

> An entity populates the attributes of the blueprint with real values.

Reviewer: ```suggestion
An entity is a concrete instance of a blueprint.   
Say you have a blueprint for a generic ship, you may then create mulltiple entities from that blueprint. All representing a single ship out in the real world.
```

| extends       | string                            | A list of references to other blueprints. The blueprint will inherit attributes from the referenced blueprints. |
| attributes    | `system/SIMOS/BlueprintAttribute` | List of attributes.                                                                                             | 

### system/SIMOS/BlueprintAttribute

Has the following attributes, where the `name`, `attributeType` and `type` is required.  

| Attribute        | Type          | Description                                                                                                                                                                                                |
|------------------|---------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| type             | string        | A reference to the BlueprintAttribute type.                                                                                                                                                                |

> | attributeType    | string        | The attribute type. Can be a reference to another blueprint, or a primitive type (string, integer, number, boolean).                                                                                       |

Reviewer 3: Mention "object"?

| dimensions       | string        | Degree of dimensions. Format; A list of integers, each integer defines the size of a dimension. "*"(wildcard) has the special meaning of "unfixed size                                                     |
| optional         | boolean       | If the attribute is optional or required                                                                                                                                                                   |
| default          | any json      | Default value for the attribute. Same type as the attribute itself                                                                                                                                         |
| enumType         | string        | A reference to an Enum entity. Value should hold the chosen enum value in the entity                                                                                                                       |

## References 

It is possible to reference other blueprint from the `type`, `extends`, `attributeType`, and `_blueprintPath_` attributes.

> ### Syntax

Reviewer 2: discussed in meeting today.
Syntax section might be better to put under guide/domain modelling instead

There are several ways to describe a reference to other blueprints. 

All of these references are understood and resolved during import. 

*By URI* - Full absolute path prefixed with protocol

```
dmss://datasource/package/entity
dmss://datasource/package/subfolder/entity
```

*By Alias* - Require dependencies to be defined somewhere in the source tree using [_meta_](./../concepts/meta.md) information.

```
ALIAS:packge/entity
ALIAS:entity
```

> *By data source* - Relative from the destination data source.

Reviewer: ```suggestion
*By data source* - Relative from the source data source.
```

---
sidebar_position: 1
---

# Blueprints and entities

Blueprints can be used to create domain models. 

:::info

A blueprint is a model and an instance of a blueprint is called an entity. 

:::

## What is a Blueprint?

A blueprint is described in [JSON](https://www.json.org/json-en.html). 

> It is an JSON object that maps keys to values.

Reviewer 3: This just explains a JSON object, so I suggest removing it
```suggestion
```

Author: Agree, I also think it is not needed. I removed it.

Below is a short version of a Car blueprint. 

```json
{
  "name": "Car",
  "type": "dmss://system/SIMOS/Blueprint",
  "description": "This describes how a car looks like",
  "attributes": [
    {
      "attributeType": "string",
      "type": "dmss://system/SIMOS/BlueprintAttribute",
      "name": "name"
    },
    {
      "attributeType": "string",
      "type": "dmss://system/SIMOS/BlueprintAttribute",
      "name": "description"
    }
  ]
}
```
Any blueprint can be understood by simply looking at it. 

> In the example above, you can find out, by looking at its attributes: `name`, `type` and `attributes`. 
> 
> * That it represent a Car model 

Reviewer 3: ```suggestion
By looking at the attributes `name`, `type` and `attributes` in the example above, you can find out that:

* It represents a Car model 
```

Author: I have updated the text with the suggested changes.

* It is of type `system/SIMOS/Blueprint`
* It has the two attributes `name` and `description`.

## How to create a Blueprint?

> The `name` should reflects what the model is.

Reviewer 3: ```suggestion
The `name` should reflect what the model is.
```
or remove it entirely

Author: I have updated the text with the suggested changes.

| Attribute        | Type                              | Description                                                                                            |
|------------------|-----------------------------------|--------------------------------------------------------------------------------------------------------|
| type             | string                            | A reference to the Blueprint type.                                                                     |
 | name             | string                            | The name of the blueprint                                                                              |
| description      | string                            | More information about the model                                                                       |
| extends          | string                            | A reference to another blueprint. The blueprint will inherit attributes from the referenced blueprint. |
| attributes       | `system/SIMOS/BlueprintAttribute` | List of attributes.                                                                                    | 

> :::info
> 
> An attribute that can be complex (reference to another blueprint) or primitive (string, integer, number, boolean). 
> 
> :::

Reviewer 3: I don't understand this. Perhaps remove it?
```suggestion
```

Author: I removed it. 

This is already described under the `attributeType` description for the `system/SIMOS/BlueprintAttribute`. Basically an attribute can either be a reference to another blueprint or a primitive like a string, number etc.

### system/SIMOS/BlueprintAttribute

> Has the following attributes, where the `attributeType` and `type` is required.  

Reviewer 3: Isn't name required?

Author: Yes it is, so I added it.

| name          | string   | Name of the attribute.                                                                                                                                                                                     |
| contained     | boolean  | Describes the conceptual belonging of one entity to another. In other words, if an entity must always be considered in the context of it's "parent" (true), or the entity has meaning on it's own (false). |
| dimensions    | string   | Degree of dimensions. Format; A list of integers, each integer defines the size of a dimension. "*"(wildcard) has the special meaning of "unfixed size                                                     |
| optional      | boolean  | If the attribute is optional or required                                                                                                                                                                   |
| default       | any json | Default value for the attribute. Same type as the attribute itself                                                                                                                                         |

> | enumType      | string   | A reference to an Enum entity. Value should hold the chosen enum value in the entity                                                                                                                       |

Reviewer 3: I'm missing type, description and label from this table

Author: Added type and description, but not label. Are label needed anymore, maybe we should try to remove it.

## References 

> It is possible to reference other blueprint from the `type`, `extends`, `attributeType`, and `_blueprintPath_` attributes.

Reviewer 3: I suppose you can only reference one blueprint? Or can you for instance extend multiple blueprints?
```suggestion
It is possible to reference another blueprint from the `type`, `extends`, `attributeType`, and `_blueprintPath_` attributes.
```

Author: Extends is an array of references [reference1, reference2, ...]. So it's possible to extend from multiple blueprints.

### Syntax

There are several ways to describe a reference to other blueprints. 

All of these references are understood and resolved during import. 

*By URI* - Full absolute path prefixed with protocol

```
dmss://datasource/package/entity
dmss://datasource/package/subfolder/entity
```

*By Alias* - Require dependencies to be defined somewhere in the source tree using [_meta_](./../concepts/meta.md) information.

```
ALIAS:packge/entity
ALIAS:entity
```

*By data source* - Relative from the destination data source.

```
/package/entity
/package/subfolder/entity
```

*By package* - Relative from the source package

```
entity
subfolder/entity
```

*By dotted* - Relative from the file (UNIX directory traversal)

```
./../entity
../subfolder/entity
../../subfolder/entity
```

## Entity

An entity populates the attributes of the blueprint with real values.

An entity of the `Car` blueprint can look like this. 

```json
{
   "type": "dmss://Car",
   "name": "Volvo",
   "description": "This is an typically car type used in Norway"
}
```

> By looking at the `type` keyword, we can see that it's of type `Car`.

Reviewer 3: I think this should be moved up, so that it is positioned right after What is a Blueprint?, since it helps understanding what a blueprint is.

Author: Good idea, I moved it.

[Blueprint]: https://github.com/equinor/data-modelling-storage-service/blob/master/src/home/system/SIMOS/Blueprint.json

> [BlueprintAttribute]: https://github.com/equinor/data-modelling-storage-service/blob/master/src/home/system/SIMOS/BlueprintAttribute.json

Reviewer 3: I didn't know you could do this. That's handy!
